### ISSUE TITLE
Parse line numbers

### ISSUE THREAD

**Issue by choldgraf** [lepture/mistune#218]
Is it possible to retrieve information about the line number of block-level elements in the AST? We are looking into using Mistune for a project and this would be quite helpful.

E.g., I'd like something like:

```python
markdown = mistune.create_markdown(renderer=mistune.AstRenderer())
markdown("# Heading 1\n\n## Heading2")
```
to return something like

```python
[{'type': 'heading',
  'children': [{'type': 'text', 'text': 'Heading 1'}],
  'level': 1,
  'lineno': 0},
 {'type': 'heading',
  'children': [{'type': 'text', 'text': 'Heading2'}],
  'level': 2,
  'lineno': 2}]
```

**Comment by chrisjsewell** [comment]
Yes ideally the 'range' (start line/char, end line/char) of each element, inline as well

**Comment by lepture** [comment]
I'm not sure if it is possible to add line numbers in AstRenderer. What is the usage of line numbers in your case?

**Comment by choldgraf** [comment]
Two quick use-cases:

* We're considering using this as a bridge to the docutils AST (what Sphinx uses), and for that it'd be helpful to have the position of elements in the source file in order to create links back to them
* We're considering building a language server protocol as people write in markdown (to provide hints and rich metadata while editing), and the tool that parses the markdown file in order to determine its structure will need to know the line number and position of every element

**Comment by araspik** [comment]
Another use-case:

* Need to be able to insert location information to checkboxes in lists so that they can be changed on-the-fly from a browser and the server can find and toggle the corresponding checkbox in the markdown source.

**Comment by ddevault** [comment]
Clarifying @araspik's use-case:

- [ ] Unchecked
- [x] Checked

In order to support this GitHub extension, we need to be able to identify the location in the Markdown source which needs to be updated to commit the change when the user toggles the checkbox in the browser. (Anyone with write access to this repo should be able to play with the checkboxes in my comment to get an idea of how it works).

**Comment by waylan** [comment]
> What is the usage of line numbers in your case?

I would be using this for a linter like tool. Perhaps a spellchecker or a link checker. In the care of a spellchecker, the tool should skip anything within code blocks or spans, so we can't just pass the entire document into a spellchecker. In the case of a link checker, we need to find all links (external, internal, absolute, relative). In each case, the easiest way to ensure we are checking the correct content is by running the document through a Markdown parser.

Of course, we could pass the rendered HTML into any number of existing tools, but that will present any errors found with the location in the HTML. However, ideally, the location (line and column) in the Markdown source document would be preferred. That way the user can easily find the error and fix it.

To implement such a tool, I would create my own custom renderer which would step through the AST, checking the content of the various nodes. For any nodes without errors, the renderer would return nothing. However, when an error is encountered, the renderer would report the error, ideally with the line and column.

See the oversimplified pseudo example of a link checker below

```python
class LinkChecker(BaseRenderer):
    NAME = 'linkchecker'
    IS_TREE = False

    def text(self, text):
        return ''

    def link(self, link, pos, children=None, title=None):
        result = validate(link):
        if result:
            return f'ERROR at {pos.line}:{pos.col} {link} --> {result.code}\n'
         return ''
```

**Comment by xanok** [comment]
> 
> See the oversimplified pseudo example of a link checker below
> 
> ```python
> class LinkChecker(BaseRenderer):
>     NAME = 'linkchecker'
>     IS_TREE = False
> 
>     def text(self, text):
>         return ''
> 
>     def link(self, link, pos, children=None, title=None):
>         result = validate(link):
>         if result:
>             return f'ERROR at {pos.line}:{pos.col} {link} --> {result.code}\n'
>          return ''
> ```

I think it is a great idea. But I would prefer a "token" parameter instead of "pos", we could add more information about the token like the "whole token text"

**Comment by chrisjsewell** [comment]
Note that, I gave up waiting on this lol, and implemented https://github.com/executablebooks/markdown-it-py instead which at least captures line numbers for blocks

**Comment by djmattyg007** [comment]
I'd like to add my support for returning line numbers. This would be immensely helpful for creating linting tools.

**Comment by ddevault** [comment]
Tracking line numbers would definitely not make mistune slower unless it the feature was egregiously poorly implemented.
